### DISCUSSION TITLE
`not` API changed on `v1.0.0-alpha.4`

### DISCUSSION THREAD

**Discussion by soywod** [zesterer/chumsky#523] [Q&A] [answered]
Hi,

I wanted to try the `v1.0.0-alpha.4` (I come from the `v0.9.2`) and I noticed that `.not()` is not acting the same. It used to return the current output (let's say `Vec<char>`) but now it seems to returns the unit. How can I have the same behaviour as before?

```rust
pub(crate) fn text_plain_part<'a>() -> impl Parser<'a, &'a str, String, extra::Err<Rich<'a, char>>>
{
    choice((
        just(SINGLE_PART_BEGIN),
        just(SINGLE_PART_END),
        just(MULTI_PART_BEGIN),
        just(MULTI_PART_END),
    ))
    // FIXME: not() returns unit instead
    .not()
    .repeated()
    .at_least(1)
    .collect::<String>()
}
```

Thank you!

**Comment by Zij-IT** [comment, marked as answer]
In this case to get the desired behavior you use something like `(map_)slice`:

```rust
pub(crate) fn text_plain_part<'a>() -> impl Parser<'a, &'a str, String, extra::Err<Rich<'a, char>>> {
    any()
        .and_is(
            choice((
                just(SINGLE_PART_BEGIN),
                just(SINGLE_PART_END),
                just(MULTI_PART_BEGIN),
                just(MULTI_PART_END),
            ))
            .not()
        )
        .repeated()
        .at_least(1)
        .map_slice(ToString::to_string)
}
```

The `any().and_is(/* snip */)` is required because an empty parse is valid for `something.not()` and therefore will infinite loop without progress when used in combination with `repeated`. By using `any().and_is` I am forcing that at least one token be parsed per iteration of "repeated"

**Reply by soywod** [reply]
It worked like a charm, thank you. I just needed to replace `.map_slice(ToString::to_string)` by `.collect::<String>()`.

**Reply by zesterer** [reply]
Sorry about this change, I'll be documenting this in the eventual migration guide. The new behaviour is strictly more general than the old behaviour, enabling a bunch of more interesting patterns, but obviously makes this specific case a little bit more verbose.

**Reply by zesterer** [reply]
>  I just needed to replace .map_slice(ToString::to_string) by .collect::<String>().

As an additional point: the new zero-copy capabilities in 1.0 allow you to avoid allocating in these circumstances by just using `.slice()` and having the parser return `&'a str` as its output type.

Even if you still want to return a `String`, `.map_slice(ToString::to_string)` will be more efficient since it's directly converting between a `&str` to a `String`, avoiding the reallocations that would be needed when collecting up individual `char`s into a `String` with the old approach.

**Reply by soywod** [reply]
Don't worry, I am aware that it is an alpha, I just wanted to try it out. Thank you for the tips!
